File: scripts/novel_CDS_validated_by_RiboSeq.py
from src.utils import read_gtf
import polars as pl

riboseq = read_gtf("nextflow_results/V47/orfanage/UCSC_tracks/riboseq.gtf")

novel_CDS = read_gtf("nextflow_results/V47/orfanage/UCSC_tracks/novel_CDS.gtf")

classification = pl.read_parquet("nextflow_results/V47/final_classification.parquet")

peptides = read_gtf("nextflow_results/V47/orfanage/UCSC_tracks/detected_peptides.gtf")

plus_strand = novel_CDS\
    .filter(pl.col("strand") == "+")\
    .sort("seqname", "start")\
    .group_by("transcript_id", maintain_order=True)\
    .agg(
        pl.col("start").first(),
        pl.col("end").first(),
        pl.col("strand").first(),
        pl.col("seqname").first(),
    )\
    .with_columns(start = pl.col("start")-1)\
    .join(
        riboseq["seqname", "start", "strand", "source"],
        how="left",
        on=["seqname", "start", "strand"],
    )\
    .filter(pl.col("source").is_not_null())\
    .unique("start")\
    .select("transcript_id", "strand", "seqname", "start", "end")

minus_strand = novel_CDS\
    .filter(pl.col("strand") == "-")\
    .sort("seqname", "end", descending=True)\
    .group_by("transcript_id", maintain_order=True)\
    .agg(
        pl.col("start").first(),
        pl.col("end").first(),
        pl.col("strand").first(),
        pl.col("seqname").first(),
    )\
    .join(
        riboseq["seqname", "end", "strand", "source"],
        how="left",
        on=["seqname", "end", "strand"],
    )\
    .filter(pl.col("source").is_not_null())\
    .unique("end")\
    .select("transcript_id", "strand", "seqname", "start", "end")

validated_ncORFs = pl.concat([plus_strand, minus_strand])

validated_ncORFs.write_csv("validated_ncORFs.txt", separator="\t")

validated_ncORFs\
    .join(
        classification[["isoform", "associated_gene"]].rename({"isoform": "transcript_id"}),
        on="transcript_id",
        how="left"
    )

File: scripts/novel_CDS_validated_by_RiboSeq.R
library(GenomicRanges)
library(dplyr)
library(readr)
library(rtracklayer)
library(arrow)

detected_peptides <- import("nextflow_results/V47/orfanage/UCSC_tracks/detected_peptides.gtf")
tx_classification <- read_parquet("nextflow_results/V47/final_classification.parquet")
protein_classification <- read_tsv("nextflow_results/V47/orfanage/SFARI.protein_classification.tsv")

validated_ncORFs <- read_delim("validated_ncORFs.txt", delim = "\t") %>%
    left_join(
        tx_classification %>% select(c(isoform, associated_gene)),
        join_by(transcript_id == isoform)
    ) %>%
    left_join(
        protein_classification %>% select(c(pb, tx_cat, protein_classification_base)),
        join_by(transcript_id == pb)
    )

validated_ncORFs %>% distinct(associated_gene)

validated_ncORFs %>% write_csv("export/validated_ncORFs.csv")

validated_ncORFs <- GRanges(
    seqnames = validated_ncORFs$seqname,
    ranges = IRanges(start = validated_ncORFs$start, end = validated_ncORFs$end),
    strand = validated_ncORFs$strand,
    transcript_id = validated_ncORFs$transcript_id
)

validated_ncORFs[queryHits(findOverlaps(validated_ncORFs, detected_peptides))]

# Number of unique peptides identified
length(unique(detected_peptides$transcript_id))

# Number of unique genes with detected peptides
length(unique(detected_peptides$gene_name))